File: env/DataType.cpp
#include "DataType.h"
#include <iostream>

using namespace std;

DataType::DataType(Type type) {
    this->type = type;
    this->length = 0;
}

DataType::DataType(DataType::Type type, int length) {
    this->type = type;
    this->length = length;
}

DataType::Type DataType::get_typecode() {
    return this->type;
}

int DataType::get_length() {
    return this->length;
}

int DataType::equals(DataType *that) {
    if (this->get_typecode() != that->get_typecode())
        return 0;

    if (this->get_typecode() == DataType::INT)
        return 1;

    if (this->get_typecode() == DataType::DOUBLE)
        return 1;

    if (this->get_typecode() == DataType::FLOAT)
        return 1;

    if (this->get_typecode() == DataType::CHAR)
        return this->get_length() == that->get_length();

    return 0;
}

string DataType::str(string name) {
    if (this->get_typecode() == DataType::INT)
        return "int32_t " + name;

    if (this->get_typecode() == DataType::DOUBLE)
        return "double " + name;

    if (this->get_typecode() == DataType::FLOAT)
        return "float " + name;

    if (this->get_typecode() == DataType::CHAR) {
        string s("char " + name + "[" + to_string(this->get_length()) + "]");
        return s;
    }

    return "";
}

string DataType::str_param_for_select(string name) {
    if (this->get_typecode() == DataType::INT)
        return "int32_t p_" + name;

    if (this->get_typecode() == DataType::DOUBLE)
        return "double p_" + name;

    if (this->get_typecode() == DataType::FLOAT)
        return "float p_" + name;

    if (this->get_typecode() == DataType::CHAR) {
        string s("const char* p_" + name);//+ "[" + to_string(this->get_length()) + "]");
        return s;
    }

    return "";
}

string DataType::str_column_for_select(string name) {
    if (this->get_typecode() == DataType::INT)
        return "int32_t c_" + name;

    if (this->get_typecode() == DataType::DOUBLE)
        return "double c_" + name;

    if (this->get_typecode() == DataType::FLOAT)
        return "float c_" + name;

    if (this->get_typecode() == DataType::CHAR) {
        string s("const char* c_" + name);//[" + to_string(this->get_length()) + "]");
        return s;
    }

    return "";
}


string DataType::str_out(string name) {
    if (this->get_typecode() == DataType::INT)
        return "int32_t " + name;

    if (this->get_typecode() == DataType::DOUBLE)
        return "double " + name;

    if (this->get_typecode() == DataType::FLOAT)
        return "float " + name;

    if (this->get_typecode() == DataType::CHAR) {
        string s("char " + name + "[" + to_string(this->get_length() + 1) + "]");
        return s;
    }

    return "";
}

string DataType::signature(string name) {
    if (this->get_typecode() == DataType::INT)
        return "int32_t " + name;

    if (this->get_typecode() == DataType::DOUBLE)
        return "double " + name;

    if (this->get_typecode() == DataType::FLOAT)
        return "float " + name;

    if (this->get_typecode() == DataType::CHAR) {
        return "const char* " + name;
    }

    return "";
}

string DataType::get_format_specifier() {
    if (this->get_typecode() == DataType::INT)
        return "%d";

    if (this->get_typecode() == DataType::DOUBLE)
        return "%f";

    if (this->get_typecode() == DataType::FLOAT)
        return "%f";

    if (this->get_typecode() == DataType::CHAR) {
        return "\\\"%s\\\"";
    }

    return "";
}

string DataType::scan_expr(string column_name) {
    if (this->get_typecode() == DataType::INT)
        return "sscanf(token, \"%d\", &(arg->" + column_name + "));";

    if (this->get_typecode() == DataType::DOUBLE)
        return "sscanf(token, \"%lf\", &(arg->" + column_name + "));";

    if (this->get_typecode() == DataType::FLOAT)
        return "sscanf(token, \"%f\", &(arg->" + column_name + "));";

    if (this->get_typecode() == DataType::CHAR) {
        return "memcpy(arg->" + column_name + ", token + sizeof(char), strlen(token)-2);";
    }

    return "";
}

string DataType::init_expr(string column_name) {
    if (this->get_typecode() == DataType::INT)
        return "new->" + column_name + " = 0;";

    if (this->get_typecode() == DataType::DOUBLE)
        return "new->" + column_name + " = 0.0;";

    if (this->get_typecode() == DataType::FLOAT)
        return "new->" + column_name + " = 0.0;";

    if (this->get_typecode() == DataType::CHAR) {
        return "memset(new->" + column_name + ", 0, " + to_string(this->get_length()) + ");";
    }

    return "";
}

string DataType::select_expr(std::string param, std::string column) {
    if (this->get_typecode() == DataType::INT)
        return "inserted->" + param + " = c_" + column + ";";

    if (this->get_typecode() == DataType::DOUBLE)
        return "inserted->" + param + " = c_" + column + ";";

    if (this->get_typecode() == DataType::FLOAT)
        return "inserted->" + param + " = c_" + column + ";";

    if (this->get_typecode() == DataType::CHAR) {
        return "memcpy(inserted->" + column + ", c_" + param + ", " + to_string(this->get_length()) +
               "); inserted->" + column + "[" + to_string(this->get_length()) + "] = '\\0';";
    }

    return "";
}

string DataType::compare_less_expr(string s1, string col1, string s2, string col2) {
    if (this->get_typecode() == DataType::INT)
        return "(" + s1 + "->" + col1 + " < " + s2 + "->" + col2 + ")";

    if (this->get_typecode() == DataType::DOUBLE)
        return "(" + s1 + "->" + col1 + " < " + s2 + "->" + col2 + ")";

    if (this->get_typecode() == DataType::FLOAT)
        return "(" + s1 + "->" + col1 + " < " + s2 + "->" + col2 + ")";

    if (this->get_typecode() == DataType::CHAR) {
        return "strncmp(" + s1 + "->" + col1 + ", " + s2 + "->" + col2 + ", " + to_string(this->get_length()) + ") < 0";
    }

    return "";
}

string DataType::compare_greater_expr(string s1, string col1, string s2, string col2) {
    if (this->get_typecode() == DataType::INT)
        return "(" + s1 + "->" + col1 + " > " + s2 + "->" + col2 + ")";

    if (this->get_typecode() == DataType::DOUBLE)
        return "(" + s1 + "->" + col1 + " > " + s2 + "->" + col2 + ")";

    if (this->get_typecode() == DataType::FLOAT)
        return "(" + s1 + "->" + col1 + " > " + s2 + "->" + col2 + ")";

    if (this->get_typecode() == DataType::CHAR) {
        return "strncmp(" + s1 + "->" + col1 + ", " + s2 + "->" + col2 + ", " + to_string(this->get_length()) + ") > 0";
    }

    return "";
}

File: pymilldb/context/DataType.py
import logging
import abc

logger = logging.getLogger('DataType')


class BaseType(abc.ABC):
    type_name = ''

    @abc.abstractmethod
    def variable(self, name, prefix=''):
        pass

    @abc.abstractmethod
    def str_param_for_select(self, name):
        pass

    @abc.abstractmethod
    def str_column_for_select(self, name):
        pass

    @abc.abstractmethod
    def str_out(self, name):
        pass

    @abc.abstractmethod
    def signature(self, name):
        pass

    @abc.abstractmethod
    def select_expr(self, param, column):
        pass

    @abc.abstractmethod
    def compare_less_expr(self, s1, col1, s2, col2):
        pass

    @abc.abstractmethod
    def compare_greater_expr(self, s1, col1, s2, col2):
        pass

    def __eq__(self, other):
        if self.type_name != other.type_name:
            return False
        else:
            if isinstance(self, Char):
                return self.size == other.size
            return True

    def __ne__(self, other):
        return not self.__eq__(other)


class ConstType(BaseType):
    @classmethod
    def variable(cls, name, prefix=''):
        return f'{cls.type_name} {prefix}{name}'

    @classmethod
    def str_param_for_select(cls, name):
        return cls.signature(name, 'p_')

    @classmethod
    def str_column_for_select(cls, name):
        return cls.signature(name, 'c_')

    @classmethod
    def str_out(cls, name):
        return cls.variable(name)

    @classmethod
    def signature(cls, name, prefix=''):
        return cls.variable(name, prefix)

    @classmethod
    def select_expr(cls, param, column):
        return f'inserted->{param} = c_{column};'

    @classmethod
    def cmp(cls, s1, col1, s2, col2, cmp):
        return f'({s1}->{col1} {cmp} {s2}->{col2})'

    @classmethod
    def compare_less_expr(cls, s1, col1, s2, col2):
        return cls.cmp(s1, col1, s2, col2, '<')

    @classmethod
    def compare_greater_expr(cls, s1, col1, s2, col2):
        return cls.cmp(s1, col1, s2, col2, '>')


class Int(ConstType):
    type_name = 'int32_t'


class Float(ConstType):
    type_name = 'float'


class Double(ConstType):
    type_name = 'double'


class ArrayType(BaseType):
    def __init__(self, size):
        self.size = size

    def variable(self, name, prefix=''):
        return f'{self.type_name} {name}[{self.size}]'

    def str_param_for_select(self, name):
        return self.signature(name, 'p_')

    def str_column_for_select(self, name):
        return self.signature(name, 'c_')

    def str_out(self, name):
        return self.variable(name)

    def signature(self, name, prefix=''):
        return f'const {self.type_name}* {name}'

    @abc.abstractmethod
    def select_expr(self, param, column):
        pass

    @abc.abstractmethod
    def compare_less_expr(self, s1, col1, s2, col2):
        pass

    @abc.abstractmethod
    def compare_greater_expr(self, s1, col1, s2, col2):
        pass


class Char(ArrayType):
    type_name = 'char'

    def select_expr(self, param, column):
        return (
            f'memcpy(inserted->{param}, c_{column}, {self.size});\n'
            f'inserted->{param}[{self.size}] = \'\\0\';'
        )

    def cmp(self, s1, col1, s2, col2, cmp):
        return f'strncmp({s1}->{col1}, {s2}->{col2}, {self.size}) {cmp} 0'

    def compare_less_expr(self, s1, col1, s2, col2):
        return self.cmp(s1, col1, s2, col2, '<')

    def compare_greater_expr(self, s1, col1, s2, col2):
        return self.cmp(s1, col1, s2, col2, '>')


class Sequence(BaseType):
    pass


name_to_type = {
    'INT': {'type': Int, 'has_size': False},
    'FLOAT': {'type': Float, 'has_size': False},
    'DOUBLE': {'type': Double, 'has_size': False},
    'CHAR': {'type': Char, 'has_size': True},
}


def get_type_by_name(name: str, size=None):
    name = name.upper()
    meta = name_to_type.get(name)
    if not meta:
        logger.error('Type: %s not found. Support INT, FLOAT, DOUBLE, CHAR and TEXT', name)
        return
    kind = meta['type']
    has_size = meta['has_size']
    if has_size and size is None:
        logger.error('Type %s needs size setting', name)
        return
    elif not has_size and size is not None:
        logging.warning('Type %s does not support sizing. Size ignoring', name)
        size = None
    if has_size:
        return kind(size)
    else:
        return kind
